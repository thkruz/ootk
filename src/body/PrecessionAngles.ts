import { Radians } from '../ootk';

// / Earth precession angles _(rad)_.
export type PrecessionAngles = {
  zeta: Radians;
  theta: Radians;
  zed: Radians;
};
