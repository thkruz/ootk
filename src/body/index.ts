export { Celestial } from './Celestial';
export { Earth } from './Earth';
export { Moon } from './Moon';
export { Sun } from './Sun';
