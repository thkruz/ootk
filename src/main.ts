/**
 * @author @thkruz Theodore Kruczek
 * @description Orbital Object ToolKit (ootk) is a collection of tools for working
 * with satellites and other orbital objects.
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Many of the classes are based off of the work of @david-rc-dayton and his
 * Pious Squid library (https://github.com/david-rc-dayton/pious_squid).
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './time/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
export * from './operations/operations.js';
export { BaseObject, Sensor } from './objects/index.js';
export { Earth } from './body/index.js';
export * from './coordinate/index.js';
export * from './observation/index.js';
export * from './data/DataHandler.js';
export * from './sgp4/index.js';
export * from './objects/index.js';
export * from './interpolator/index.js';
export * from './force/index.js';
export * from './maneuver/index.js';
export * from './optimize/index.js';
export * from './orbit_determination/index.js';
export * from './body/index.js';
export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './coordinate/index.js';
export * from './operations/index.js';
export * from './force/index.js';
export * from './propagator/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
