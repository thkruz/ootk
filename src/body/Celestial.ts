/**
 * @author @thkruz Theodore Kruczek
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

import { AzEl, Degrees, Kilometers, RAD2DEG, RaDec, Radians, Sun } from '../main.js';

/**
 * Celestial is a static class that provides methods for calculating the position of celestial objects such as the Sun,
 * Moon, and planets in the sky.
 */
export class Celestial {
  private constructor() {
    // disable constructor
  }

  /**
   * Calculates the azimuth and elevation of a celestial object at a given date, latitude,
   * longitude, right ascension, and declination.
   * @param date - The date for which to calculate the azimuth and elevation.
   * @param lat - The latitude of the observer.
   * @param lon - The longitude of the observer.
   * @param ra - The right ascension of the celestial object.
   * @param dec - The declination of the celestial object.
   * @returns An object containing the azimuth and elevation in degrees.
   */
  static azEl(date: Date, lat: Degrees, lon: Degrees, ra: Radians, dec: Radians): AzEl<Degrees> {
    const c: RaDec = {
      ra,
      dec,
      dist: 0 as Kilometers,
    };
    const azEl = Sun.azEl(date, lat, lon, c);

    const el = <Radians>(azEl.el + Celestial.atmosphericRefraction(azEl.el)); // elevation correction for refraction

    return {
      az: (azEl.az * RAD2DEG) as Degrees,
      el: (el * RAD2DEG) as Degrees,
    };
  }

  /**
   * Atmospheric refraction in astronomy, refers to the bending of light as it passes through the Earth's
   * atmosphere. This effect is most noticeable for celestial objects like stars and planets when they are
   * close to the horizon. Here's a breakdown of how it works:
   *
   * Actual Position: Due to this bending of light, the apparent position of a celestial object is slightly
   * different from its true position in the sky. When a star or planet is near the horizon, the effect is more
   * pronounced because the light path passes through more of the Earth's atmosphere, which increases the amount of
   * bending.
   *
   * A familiar example of atmospheric refraction is observed during sunrise and sunset. The Sun appears to
   * be above the horizon when it is actually just below it. This is because the light from the Sun is bent
   * upwards as it passes through the atmosphere.
   * @param h - elevation
   * @returns refraction
   */
  static atmosphericRefraction(h: Radians): Radians {
    if (h < 0) {
      h = <Radians>0;
    }

    return <Radians>(0.0002967 / Math.tan(h + 0.00312536 / (h + 0.08901179)));
  }

  /**
   * Calculate the declination. Similar to latitude on Earth, declination is another celestial coordinate.
   * It measures how far north or south an object is from the celestial equator
   * @param l - ecliptic longitude
   * @param b - ecliptic latitude
   * @returns declination
   */
  static declination(l: number, b: number): Radians {
    return <Radians>Math.asin(Math.sin(b) * Math.cos(Sun.e) + Math.cos(b) * Math.sin(Sun.e) * Math.sin(l));
  }

  /**
   * Calculate the right ascension. This is a celestial coordinate used to determine the position of objects
   * in the sky. It's analogous to longitude on Earth. Right Ascension indicates how far east an object is
   * from the vernal equinox along the celestial equator.
   * @param l - ecliptic longitude
   * @param b - ecliptic latitude
   * @returns right ascension
   */
  static rightAscension(l: number, b: number): Radians {
    return <Radians>Math.atan2(Math.sin(l) * Math.cos(Sun.e) - Math.tan(b) * Math.sin(Sun.e), Math.cos(l));
  }

  /**
   * Calculate the elevation. Elevation, or altitude, is the angle between an object in the sky and the
   * observer's local horizon. It's commonly expressed in degrees, where 0 degrees is right at the horizon
   * and 90 degrees is directly overhead (the zenith), but we are using radians to support trigonometric
   * functions like Math.sin() and Math.cos().
   * @param H - siderealTime
   * @param phi - latitude
   * @param dec - The declination of the sun
   * @returns elevation
   */
  static elevation(H: number, phi: Radians, dec: Radians): Radians {
    return <Radians>Math.asin(Math.sin(phi) * Math.sin(dec) + Math.cos(phi) * Math.cos(dec) * Math.cos(H));
  }

  /**
   * Calculate the azimuth. This is a compass direction measurement. Azimuth measures the angle along
   * the horizon from a specific reference direction (usually true north) to the point where a vertical
   * line from the object intersects the horizon.
   * @param H - siderealTime
   * @param phi - latitude
   * @param dec - The declination of the sun
   * @returns azimuth in rad
   */
  static azimuth(H: number, phi: Radians, dec: Radians): Radians {
    return <Radians>(Math.PI + Math.atan2(Math.sin(H), Math.cos(H) * Math.sin(phi) - Math.tan(dec) * Math.cos(phi)));
  }
}
