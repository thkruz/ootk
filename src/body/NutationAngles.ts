/**
 * @author @thkruz Theodore Kruczek
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

import { Radians } from '../main.js';

/** Represents the nutation angles. */
export type NutationAngles = {
  /** The nutation in longitude (Δψ) in radians. */
  dPsi: Radians;
  /** The nutation in obliquity (Δε) in radians. */
  dEps: Radians;
  /** The mean obliquity of the ecliptic (ε₀) in radians. */
  mEps: Radians;
  /** The true obliquity of the ecliptic (ε) in radians. */
  eps: Radians;
  /** The equation of the equinoxes (ΔΔt) in radians. */
  eqEq: Radians;
  /** The Greenwich Apparent Sidereal Time (GAST) in radians. */
  gast: Radians;
};
