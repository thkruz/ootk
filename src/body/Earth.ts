/**
 * @author @thkruz Theodore Kruczek
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

import {
  angularDiameter,
  AngularDiameterMethod,
  asec2rad,
  DataHandler,
  DEG2RAD,
  earthGravityParam,
  EpochUTC,
  evalPoly,
  Kilometers,
  RAD2DEG,
  Radians,
  RadiansPerSecond,
  secondsPerDay,
  secondsPerSiderealDay,
  TAU,
  ttasec2rad,
  Vector3D,
} from '../main.js';
import { NutationAngles } from './NutationAngles.js';
import { PrecessionAngles } from './PrecessionAngles.js';

// / Earth metrics and operations.
export class Earth {
  private constructor() {
    // disable constructor
  }

  // / Earth gravitational parameter _(km²/s³)_.
  static readonly mu: number = earthGravityParam;

  // / Earth equatorial radius.
  static readonly radiusEquator = 6378.1363 as Kilometers;

  // / Earth coefficient of flattening _(unitless)_.
  static readonly flattening: number = 1.0 / 298.257223563;

  // / Earth polar radius.
  static readonly radiusPolar = (Earth.radiusEquator * (1.0 - Earth.flattening)) as Kilometers;

  // / Earth mean radius.
  static readonly radiusMean = ((2.0 * Earth.radiusEquator + Earth.radiusPolar) / 3.0) as Kilometers;

  // / Earth eccentricity squared _(unitless)_.
  static readonly eccentricitySquared: number = Earth.flattening * (2.0 - Earth.flattening);

  // / Earth J2 effect coefficient _(unitless)_.
  static readonly j2: number = 1.08262668355315e-3;

  // / Earth J3 effect coefficient _(unitless)_.
  static readonly j3: number = -2.53265648533224e-6;

  // / Earth J4 effect coefficient _(unitless)_.
  static readonly j4: number = -1.619621591367e-6;

  // / Earth J5 effect coefficient _(unitless)_.
  static readonly j5: number = -2.27296082868698e-7;

  // / Earth J6 effect coefficient _(unitless)_.
  static readonly j6: number = 5.40681239107085e-7;

  // / Earth rotation vector _(rad/s)_.
  static readonly rotation = new Vector3D<RadiansPerSecond>(
    0 as RadiansPerSecond,
    0 as RadiansPerSecond,
    7.292115146706979e-5 as RadiansPerSecond,
  );

  // / Calculate mean motion _(rad/s)_ from a given [semimajorAxis] _(km)_.
  static smaToMeanMotion(semimajorAxis: Kilometers): RadiansPerSecond {
    return Math.sqrt(Earth.mu / (semimajorAxis * semimajorAxis * semimajorAxis)) as RadiansPerSecond;
  }

  /**
   * Converts revolutions per day to semi-major axis.
   * @param rpd - The number of revolutions per day.
   * @returns The semi-major axis value.
   */
  static revsPerDayToSma(rpd: number): number {
    return Earth.mu ** (1 / 3) / ((TAU * rpd) / secondsPerDay) ** (2 / 3);
  }

  // / Calculate Earth [PrecessionAngles] at a given UTC [epoch].
  static precession(epoch: EpochUTC): PrecessionAngles {
    const t = epoch.toTT().toJulianCenturies();
    const zeta = evalPoly(t, Earth.zetaPoly_);
    const theta = evalPoly(t, Earth.thetaPoly_);
    const zed = evalPoly(t, Earth.zedPoly_);

    return { zeta: zeta as Radians, theta: theta as Radians, zed: zed as Radians };
  }

  // / Calculate Earth [NutationAngles] for a given UTC [epoch].
  static nutation(epoch: EpochUTC): NutationAngles {
    const t = epoch.toTT().toJulianCenturies();
    const moonAnom = evalPoly(t, Earth.moonAnomPoly_);
    const sunAnom = evalPoly(t, Earth.sunAnomPoly_);
    const moonLat = evalPoly(t, Earth.moonLatPoly_);
    const sunElong = evalPoly(t, Earth.sunElongPoly_);
    const moonRaan = evalPoly(t, Earth.moonRaanPoly_);
    let deltaPsi = 0.0;
    let deltaEpsilon = 0.0;
    const dh = DataHandler.getInstance();

    for (let i = 0; i < 4; i++) {
      const [a1, a2, a3, a4, a5, ai, bi, ci, di] = dh.getIau1980Coeffs(i);
      const arg = a1 * moonAnom + a2 * sunAnom + a3 * moonLat + a4 * sunElong + a5 * moonRaan;
      const sinC = ai + bi * t;
      const cosC = ci + di * t;

      deltaPsi += sinC * Math.sin(arg);
      deltaEpsilon += cosC * Math.cos(arg);
    }
    deltaPsi *= ttasec2rad;
    deltaEpsilon *= ttasec2rad;
    const meanEpsilon = evalPoly(t, Earth.meanEpsilonPoly_);
    const epsilon = meanEpsilon + deltaEpsilon;
    const eqEq =
      deltaPsi * Math.cos(meanEpsilon) +
      0.00264 * asec2rad * Math.sin(moonRaan) +
      0.000063 * asec2rad * Math.sin(2.0 * moonRaan);
    const gast = epoch.gmstAngle() + eqEq;

    return {
      dPsi: deltaPsi as Radians,
      dEps: deltaEpsilon as Radians,
      mEps: meanEpsilon as Radians,
      eps: epsilon as Radians,
      eqEq: eqEq as Radians,
      gast: gast as Radians,
    };
  }

  // / Convert a [semimajorAxis] _(km)_ to an eastward drift rate _(rad/day)_.
  static smaToDrift(semimajorAxis: number): number {
    const t = (TAU * Math.sqrt(semimajorAxis ** 3 / Earth.mu)) / secondsPerSiderealDay;

    return (1.0 - t) * TAU;
  }

  // / Convert a [semimajorAxis] _(km)_ to an eastward drift rate _(°/day)_.
  static smaToDriftDegrees(semimajorAxis: number): number {
    return Earth.smaToDrift(semimajorAxis) * RAD2DEG;
  }

  // / Convert an eastward [driftRate] _(rad/day)_ to a semimajor-axis _(km)_.
  static driftToSemimajorAxis(driftRate: number): number {
    const t = (-driftRate / TAU + 1) * secondsPerSiderealDay;

    return ((Earth.mu * t * t) / (4 * Math.PI * Math.PI)) ** (1 / 3);
  }

  // / Convert an eastward [driftRate] _(°/day)_ to a semimajor-axis _(km)_.
  static driftDegreesToSma(driftRate: number): number {
    return Earth.driftToSemimajorAxis(DEG2RAD * driftRate);
  }

  /**
   * Calculates the diameter of the Earth based on the satellite position.
   * @param satPos The position of the satellite.
   * @returns The diameter of the Earth.
   */
  static diameter(satPos: Vector3D): number {
    return angularDiameter(Earth.radiusEquator * 2, satPos.magnitude(), AngularDiameterMethod.Sphere);
  }

  // / Earth precession `zeta` polynomial coefficients.
  private static zetaPoly_: Float64Array = Float64Array.from([
    0.017998 * asec2rad,
    0.30188 * asec2rad,
    2306.2181 * asec2rad,
    0.0,
  ]);

  // / Earth precession `theta` polynomial coefficients.
  private static thetaPoly_: Float64Array = Float64Array.from([
    -0.041833 * asec2rad,
    -0.42665 * asec2rad,
    2004.3109 * asec2rad,
    0.0,
  ]);

  // / Earth precession `zed` polynomial coefficients.
  private static zedPoly_: Float64Array = Float64Array.from([
    0.018203 * asec2rad,
    1.09468 * asec2rad,
    2306.2181 * asec2rad,
    0,
  ]);

  private static moonAnomPoly_: Float64Array = Float64Array.from([
    1.4343e-5 * DEG2RAD,
    0.0088553 * DEG2RAD,
    (1325.0 * 360.0 + 198.8675605) * DEG2RAD,
    134.96340251 * DEG2RAD,
  ]);

  // / Earth nutation Sun anomaly polynomial coefficients.
  private static sunAnomPoly_: Float64Array = Float64Array.from([
    3.8e-8 * DEG2RAD,
    -0.0001537 * DEG2RAD,
    (99.0 * 360.0 + 359.0502911) * DEG2RAD,
    357.52910918 * DEG2RAD,
  ]);

  // / Earth nutation Moon latitude polynomial coefficients.
  private static moonLatPoly_: Float64Array = Float64Array.from([
    -2.88e-7 * DEG2RAD,
    -0.003542 * DEG2RAD,
    (1342.0 * 360.0 + 82.0174577) * DEG2RAD,
    93.27209062 * DEG2RAD,
  ]);

  // / Earth nutation Sun elongation polynomial coefficients.
  private static sunElongPoly_: Float64Array = Float64Array.from([
    1.831e-6 * DEG2RAD,
    -0.0017696 * DEG2RAD,
    (1236.0 * 360.0 + 307.1114469) * DEG2RAD,
    297.85019547 * DEG2RAD,
  ]);

  // / Earth nutation Moon right-ascension polynomial coefficients.
  private static moonRaanPoly_: Float64Array = Float64Array.from([
    2.139e-6 * DEG2RAD,
    0.0020756 * DEG2RAD,
    -(5.0 * 360.0 + 134.1361851) * DEG2RAD,
    125.04455501 * DEG2RAD,
  ]);

  // / Earth nutation mean epsilon polynomial coefficients.
  private static meanEpsilonPoly_: Float64Array = Float64Array.from([
    0.001813 * asec2rad,
    -0.00059 * asec2rad,
    -46.815 * asec2rad,
    84381.448 * asec2rad,
  ]);
}
