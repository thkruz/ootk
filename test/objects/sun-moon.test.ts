import { Celestial } from '../../src/body/Celestial';
import { Moon } from '../../src/body/Moon';
import { Sun } from '../../src/body/Sun';
import { Degrees, Meters, SunTime } from '../../src/main';

/*
 * Use number of milliseconds since epoch instead of local year, month, day, etc
 * for consistency across machines
 */
const dateObj = new Date(1661406000000);

describe('Sun and Moon', () => {
  test('Sun Unit Tests', () => {
    const d = Sun.julian2date(Sun.date2jSince2000(dateObj));
    const c = Sun.raDec(d);

    Celestial.azEl(dateObj, 0 as Degrees, 0 as Degrees, c.ra, c.dec);
    Sun.azEl(dateObj, 0 as Degrees, 0 as Degrees);
  });

  test('Moon Unit Tests', () => {
    expect(Moon.getMoonIllumination(dateObj)).toMatchSnapshot();

    Moon.rae(dateObj, 0 as Degrees, 0 as Degrees);

    /*
     * TODO: Fix this test to work on ci/cd pipeline
     * Moon.getMoonTimes(dateObj, 0 as Degrees, 0 as Degrees, true);
     * Moon.getMoonTimes(dateObj, -10 as Degrees, -10 as Degrees, false);
     */
  });

  test('getMoonIllumination returns fraction and angle of moons illuminated limb and phase', () => {
    const moonIllum = Moon.getMoonIllumination(dateObj);

    expect(moonIllum).toMatchSnapshot();
  });
});

/**
 * @author Robert Myers @xqjibz See: https://github.com/mourner/suncalc/pull/35
 */
describe('Test for #6 fix', () => {
  test('Test for #6 fix', () => {
    const date = new Date('2013-03-05UTC');
    const result = (Sun.azEl(date, 50.5 as Degrees, 30.5 as Degrees).az * 180) / Math.PI;

    expect(result).toBeCloseTo(36.742354609606814);
  });
});

describe('Test for variety of date/time stamps', () => {
  describe('getTimes day detection works with a variety of date times', () => {
    const lat = 47.606209;
    const lng = -122.332069;
    // const testDateDay = 4;
    const testDateStrings = [
      'Mon, 04 Mar 2013 00:00:01 UTC+2',
      'Mon, 04 Mar 2013 00:00:01 PDT',
      'Mon, 04 Mar 2013 00:00:01 PST',
      'Mon, 04 Mar 2013 00:00:01 EDT',
      'Mon, 04 Mar 2013 00:00:01 EST',
      'Mon, 04 Mar 2013 00:00:01 UTC',
      'Mon, 04 Mar 2013 12:00:00 PDT',
      'Mon, 04 Mar 2013 12:00:00 PST',
      'Mon, 04 Mar 2013 12:00:00 EDT',
      'Mon, 04 Mar 2013 12:00:00 EST',
      'Mon, 04 Mar 2013 12:00:00 UTC',
      'Mon, 04 Mar 2013 23:59:59 PDT',
      'Mon, 04 Mar 2013 23:59:59 PST',
      'Mon, 04 Mar 2013 23:59:59 EDT',
      'Mon, 04 Mar 2013 23:59:59 EST',
      'Mon, 04 Mar 2013 23:59:59 UTC',
    ];

    for (let i = 0, l = testDateStrings.length; i < l; i++) {
      test(`${testDateStrings[i]}`, () => {
        const date = new Date(testDateStrings[i] as string);
        const testDateDay = date.getDate();

        const times = Sun.getTimes(date, lat as Degrees, lng as Degrees);

        expect(times.solarNoon.getDate()).toEqual(testDateDay);
      });
    }
  });
});

describe('Suncalc.js tests', () => {
  const date = new Date('2013-03-05 UTC');
  const lat = 50.5;
  const lon = 30.5;
  const alt = 2000;

  const testTimes = {
    solarNoon: new Date('2013-03-05T10:10:57Z'),
    nadir: new Date('2013-03-05T22:10:57Z'),
    sunriseStart: new Date('2013-03-05T04:34:56Z'),
    sunsetEnd: new Date('2013-03-05T15:46:57Z'),
    sunriseEnd: new Date('2013-03-05T04:38:19Z'),
    sunsetStart: new Date('2013-03-05T15:43:34Z'),
    civilDawn: new Date('2013-03-05T04:02:17Z'),
    civilDusk: new Date('2013-03-05T16:19:36Z'),
    nauticalDawn: new Date('2013-03-05T03:24:31Z'),
    nauticalDusk: new Date('2013-03-05T16:57:22Z'),
    astronomicalDawn: new Date('2013-03-05T02:46:17Z'),
    astronomicalDusk: new Date('2013-03-05T17:35:36Z'),
    goldenHourDawnEnd: new Date('2013-03-05T05:19:01Z'),
    goldenHourDuskStart: new Date('2013-03-05T15:02:52Z'),
  };

  const heightTestTimes = {
    solarNoon: new Date('2013-03-05T10:10:57Z'),
    nadir: new Date('2013-03-05T22:10:57Z'),
    sunriseStart: new Date('2013-03-05T04:25:07Z'),
    sunsetEnd: new Date('2013-03-05T15:56:46Z'),
  };

  test('getTimes returns sun phases for the given date and location', () => {
    const times = Sun.getTimes(date, lat as Degrees, lon as Degrees, 0 as Meters, true);

    // eslint-disable-next-line guard-for-in
    for (const i in testTimes) {
      const key = i as keyof typeof testTimes;

      expect(times[key].toUTCString()).toEqual(new Date(testTimes[key]).toUTCString());
    }
  });

  test('getTimes adjusts sun phases when additionally given the observer height', () => {
    const times = Sun.getTimes(date, lat as Degrees, lon as Degrees, alt as Meters, true);

    // eslint-disable-next-line guard-for-in
    for (const i in heightTestTimes) {
      const key = i as keyof typeof heightTestTimes;

      expect(times[key].toUTCString()).toEqual(new Date(heightTestTimes[key]).toUTCString());
    }
  });

  test('getSunAzEl returns azimuth and altitude for the given time and location', () => {
    const sunPos = Sun.azEl(date, lat as Degrees, lon as Degrees);

    expect(sunPos).toMatchSnapshot();
  });

  test('getMoonIllumination returns fraction and angle of moons illuminated limb and phase', () => {
    const moonIllum = Moon.getMoonIllumination(date);

    expect(moonIllum).toMatchSnapshot();
  });

  test('getMoonPosition returns moon position data given time and location', () => {
    const moonPos = Moon.rae(date, lat as Degrees, lon as Degrees);

    expect(moonPos).toMatchSnapshot();
  });

  test('getMoonTimes returns moon rise and set times', () => {
    const moonTimes = Moon.getMoonTimes(new Date('2013-03-04UTC'), lat as Degrees, lon as Degrees, true);

    expect(moonTimes.rise?.toUTCString()).toEqual('Mon, 04 Mar 2013 23:54:29 GMT');
    expect(moonTimes.set?.toUTCString()).toEqual('Mon, 04 Mar 2013 07:47:58 GMT');
  });
});

describe('Tests from Hypnos3', () => {
  const date = new Date('2013-03-05UTC');
  const lat = -34.0;
  const lon = 151.0;

  const testTimes: SunTime = {
    // southern hemisphere
    solarNoon: new Date('2013-03-05T02:09:01.832Z'),
    nadir: new Date('2013-03-05T14:09:01.832Z'),
    goldenHourDuskStart: new Date('2013-03-05T07:56:33.416Z'),
    goldenHourDawnEnd: new Date('2013-03-04T20:21:30.248Z'),
    sunsetStart: new Date('2013-03-05T08:27:05.997Z'),
    sunriseEnd: new Date('2013-03-04T19:50:57.667Z'),
    sunsetEnd: new Date('2013-03-05T08:29:41.731Z'),
    sunriseStart: new Date('2013-03-04T19:48:21.933Z'),
    goldenHourDuskEnd: new Date('2013-03-05T08:30:30.554Z'),
    goldenHourDawnStart: new Date('2013-03-04T19:47:33.110Z'),
    blueHourDuskStart: new Date('2013-03-05T08:45:10.179Z'),
    blueHourDawnEnd: new Date('2013-03-04T19:32:53.485Z'),
    civilDusk: new Date('2013-03-05T08:54:59.722Z'),
    civilDawn: new Date('2013-03-04T19:23:03.942Z'),
    blueHourDuskEnd: new Date('2013-03-05T09:04:52.263Z'),
    blueHourDawnStart: new Date('2013-03-04T19:13:11.401Z'),
    nauticalDusk: new Date('2013-03-05T09:24:48.289Z'),
    nauticalDawn: new Date('2013-03-04T18:53:15.375Z'),
    amateurDusk: new Date('2013-03-05T09:39:57.120Z'),
    amateurDawn: new Date('2013-03-04T18:38:06.544Z'),
    astronomicalDusk: new Date('2013-03-05T09:55:18.657Z'),
    astronomicalDawn: new Date('2013-03-04T18:22:45.007Z'),
  };

  test('southern hemisphere', () => {
    const times = Sun.getTimes(date, lat as Degrees, lon as Degrees);

    // eslint-disable-next-line guard-for-in
    for (const i in testTimes) {
      const key = i as keyof typeof testTimes;

      expect(times[key].toUTCString()).toEqual(new Date(testTimes[key]).toUTCString());
    }
  });
});
