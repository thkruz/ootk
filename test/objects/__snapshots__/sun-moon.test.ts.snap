// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sun and Moon Moon Unit Tests 1`] = `
Object {
  "angle": 2.0303507547114927,
  "fraction": 0.046426852648024874,
  "next": Object {
    "date": "2022-08-27T06:46:57.840Z",
    "firstQuarter": Object {
      "date": "2022-09-03T15:57:58.534Z",
      "value": 1662220678534.5,
    },
    "fullMoon": Object {
      "date": "2022-09-11T01:08:59.229Z",
      "value": 1662858539229,
    },
    "newMoon": Object {
      "date": "2022-08-27T06:46:57.840Z",
      "value": 1661582817840,
    },
    "thirdQuarter": Object {
      "date": "2022-09-18T10:19:59.923Z",
      "value": 1663496399923.5,
    },
    "type": "newMoon",
    "value": 1661582817840,
  },
  "phase": Object {
    "code": ":waning_crescent_moon:",
    "css": "wi-moon-wan-cres",
    "emoji": "ðŸŒ˜",
    "from": 0.783863193308711,
    "id": "waningCrescentMoon",
    "name": "Waning Crescent",
    "to": 0.966136806691289,
    "weight": 6.3825,
  },
  "phaseValue": 0.9308720062791223,
}
`;

exports[`Sun and Moon getMoonIllumination returns fraction and angle of moons illuminated limb and phase 1`] = `
Object {
  "angle": 2.0303507547114927,
  "fraction": 0.046426852648024874,
  "next": Object {
    "date": "2022-08-27T06:46:57.840Z",
    "firstQuarter": Object {
      "date": "2022-09-03T15:57:58.534Z",
      "value": 1662220678534.5,
    },
    "fullMoon": Object {
      "date": "2022-09-11T01:08:59.229Z",
      "value": 1662858539229,
    },
    "newMoon": Object {
      "date": "2022-08-27T06:46:57.840Z",
      "value": 1661582817840,
    },
    "thirdQuarter": Object {
      "date": "2022-09-18T10:19:59.923Z",
      "value": 1663496399923.5,
    },
    "type": "newMoon",
    "value": 1661582817840,
  },
  "phase": Object {
    "code": ":waning_crescent_moon:",
    "css": "wi-moon-wan-cres",
    "emoji": "ðŸŒ˜",
    "from": 0.783863193308711,
    "id": "waningCrescentMoon",
    "name": "Waning Crescent",
    "to": 0.966136806691289,
    "weight": 6.3825,
  },
  "phaseValue": 0.9308720062791223,
}
`;

exports[`Suncalc.js tests getMoonIllumination returns fraction and angle of moons illuminated limb and phase 1`] = `
Object {
  "angle": 1.6260163867660833,
  "fraction": 0.42576863432885875,
  "next": Object {
    "date": "2013-03-12T04:53:32.814Z",
    "firstQuarter": Object {
      "date": "2013-03-19T14:04:33.508Z",
      "value": 1363701873508.5,
    },
    "fullMoon": Object {
      "date": "2013-03-26T23:15:34.203Z",
      "value": 1364339734203,
    },
    "newMoon": Object {
      "date": "2013-03-12T04:53:32.814Z",
      "value": 1363064012814,
    },
    "thirdQuarter": Object {
      "date": "2013-04-03T08:26:34.897Z",
      "value": 1364977594897.5,
    },
    "type": "newMoon",
    "value": 1363064012814,
  },
  "phase": Object {
    "code": ":last_quarter_moon:",
    "css": "wi-moon-third-quart",
    "emoji": "ðŸŒ—",
    "from": 0.716136806691289,
    "id": "thirdQuarterMoon",
    "name": "third Quarter",
    "to": 0.783863193308711,
    "weight": 1,
  },
  "phaseValue": 0.773716250255787,
}
`;

exports[`Suncalc.js tests getMoonPosition returns moon position data given time and location 1`] = `
Object {
  "az": 4.745208978619362,
  "el": -0.4894905500271524,
  "parallacticAngle": 0.7460123936598905,
  "rng": 364139.79362404963,
}
`;

exports[`Suncalc.js tests getSunAzEl returns azimuth and altitude for the given time and location 1`] = `
Object {
  "az": 3.6915486281450507,
  "el": 0.5021704545690489,
}
`;
